### PR TITLE
setAttribute() function for defining custom shader attributes

### LINKED ISSUE #5120: p5Shader - Is there a way to specify custom attributes?
### How would this new feature help [increase access](https://github.com/processing/p5.js/blob/main/contributor_docs/access.md) to p5.js?
Potential to expand capability of p5's WebGL mode.

#### Most appropriate sub-area of p5.js?

- [x] WebGL

#### Feature enhancement details:
Is there a way to specify custom attributes when working with [shaders][0]?

From what I've gathered from [this contributor_doc][1], these are the available attributes (let me know if this assumption is incorrect):
- `attribute vec3 aPosition`
- `attribute vec3 aNormal`
- `attribute vec2 aTexCoord`
- `attribute vec4 aDirection`
- `attribute vec4 aVertexColor`

Suppose I want to pass some custom attributes to a shader. (For example, I want to specify a cube (or other custom geometry such as a 3D model or something generated dynamically) by passing in some vertices as one attribute, and some vertex colors as another attribute). How can I go about this? Is there a `setAttribute` equivalent to [`setUniform`][3]? If not, is there a way to do this using private p5 methods?

---

_Sidenote:_  
It would be great if the shader documentation on [p5js.org/reference][2] was a bit more fleshed out. For example:
- list all available attributes, and give a brief explanation of how they are generated
- give an example of how to use each attribute. (For example the use of `beginShape` and `endShape` to specify custom geometry (`aPostion`), and vertex colors (`aVertexColor`) (as seen here https://github.com/processing/p5.js/issues/4274))

---

_Related issues:_  
- https://github.com/processing/p5.js/issues/5022
- https://github.com/processing/p5.js/issues/2022

[0]: https://p5js.org/reference/#/p5/shader
[1]: https://github.com/processing/p5.js/blob/main/contributor_docs/webgl_mode_architecture.md
[2]: https://p5js.org/reference/
[3]: https://p5js.org/reference/#/p5.Shader/setUniform

### PR DESCRIPTION
**Pull request by lukeplowden**
Resolves #5120 

In response to the issue above I've added a global [`setAttribute()`](https://github.com/lukeplowden/p5-attributes/blob/3c4fa80f65758ee25f1e84e967ff24b880c65617/src/core/shape/vertex.js#L2256C1-L2260C3) function and a method of the same name on [`p5.Geometry`](https://github.com/lukeplowden/p5-attributes/blob/3c4fa80f65758ee25f1e84e967ff24b880c65617/src/webgl/p5.Geometry.js#L1921-L1935).

https://github.com/user-attachments/assets/d7a37c2c-f9a6-47b6-b78e-81d7fdd55c1f

Attributes are per vertex data declared at the top of the vertex shader like so:
```glsl
  attribute vec3 aPosition;
```
Currently, p5.js has a number of predefined attributes including `aPosition`, `aVertexColor`, `aNormal`, etc. These changes allow the user to define custom attributes in **two ways**, using an API similar to `Shader.setUniform()`. 

1. By calling the global function `setAttribute(attributeName, data)` which applies the custom attribute to subsequent calls to `vertex()`, `bezierVertex()`, or `curveVertex()`, within a `beginShape()`->`endShape()` block, i.e. on **immediate mode** geometry. This is the same behaviour as `fill()` or `normal()` for per-vertex data.
```js
  setAttribute('aCustom', [1, 0, 0]);
```
2. By calling the member function`Geometry.setAttribute()` in case the user wants to directly interface with `p5.Geometry` objects, i.e. on **retained mode** geometry. This essentially creates an array and pushes a new value on the `p5.Geometry` object, which is unlike other methods on `p5.Geometry`.
```js
  myGeometry.setAttribute('aCustom', [1, 0, 0]);
```
The parameters for both functions are the same, a string `attributeName` and a number or array of numbers `data`. 

#### PR Checklist
- [x] `npm run lint` passes
- [x] [Inline documentation] is included / updated
- [x] [Unit tests] are included / updated

This is a draft pull request as I'm finishing off the tests and documentation right now. Would like to hear other WebGL contributor's thoughts on progress @perminder-17 @Garima3110 

---
## Edge cases
There are a few edge cases related to the need for attributes to be properly set for *each vertex*. If `attribute vec3 aCustom;` is declared and used in a shader, there should be as many values for `aCustom` as there are for `aPosition`.

### Global function (immediate mode)
1. The user may call `setAttribute()` *after* the first `vertex()` call. In this case `vertex()` zero-fills the previous values for each call which came before. It behaves like this:
```js
function draw(){
	shader(customAttributeShader);
	beginShape();
	vertex(-1,0,0); // This has no custom attribute set, so it is [0, 0, 0]
	setAttribute('aCustom', [1,1,1]); // Attribute set for following vertices
	vertex(1,0,0); // [1, 1, 1]
	vertex(0,1,0);// [1, 1, 1]
	endShape();
}
```
2. The user may not set any attributes but their shader uses them. We could either give a friendly error that the shader is using undefined attributes, or fill all values with zero explicitly, or a combination.  
```js
function draw(){
	shader(customAttributeShader);
	beginShape();
	vertex(-1,0,0); // no value set for aCustom
	vertex(1,0,0);
	vertex(0,1,0);
	endShape();
}
```
### Retained mode
Because the `Geometry.setAttribute(attributeName, data)` method pushes to an array `Geometry[attributeName]`:
```js
let geo;

async function setup(){
	geo = new p5.Geometry();
	let v0 = new p5.Vertex(-1,0,0);
	let v1 = new p5.Vertex(1,0,0);
	let v2 = new p5.Vertex(0,1,0);
	geo.setAttribute('aCustom', [1,0,0]); // Only one custom attribute has been set
	geo.vertices.push(v0, v1, v2); // Three vertex attributes have been set
} 
```
In this case we plan to just give a Friendly Error saying that one of the vertex attributes has not been properly filled.

### CODE REVIEW

**Comment by davepagurek** [comment]
It looks like we've got some changes from the main branch in here, do we have any conflicts if we `git rebase dev-2.0`? If it tries to replay some of the main branch commits, maybe we can skip the main branch commits by doing an interactive rebase with `git rebase -i dev-2.0` and then [deleting the lines with the unrelated commits](https://stackoverflow.com/a/12604880)?

**Comment by lukeplowden** [comment]
> It looks like we've got some changes from the main branch in here, do we have any conflicts if we `git rebase dev-2.0`? If it tries to replay some of the main branch commits, maybe we can skip the main branch commits by doing an interactive rebase with `git rebase -i dev-2.0` and then [deleting the lines with the unrelated commits](https://stackoverflow.com/a/12604880)?

Thanks for spotting that, should be good now!

File: src/webgl/p5.Geometry.js [hunk 1]
```diff
@@ -1910,6 +1917,19 @@ p5.Geometry = class Geometry {
     }
     return this;
   }
+
+  setAttribute(attributeName, data){
+    const size = data.length ? data.length : 1;
+    if (!this.hasOwnProperty(attributeName)){
```
Comment (davepagurek) [inline src/webgl/p5.Geometry.js:1923]: It's probably unlikely, but in theory a bit of user data could clash with some of our own properties, e.g. if someone tries to make their own `vertices` property. Does anything else break if we store the user attributes in `this.attribs[attributeName]`? If `RenderBuffer` doesn't handle nested properties like that, we could also always record the initial properties and throw an error if a user tries to set an attribute with a core property name.

File: src/webgl/p5.RendererGL.Immediate.js [hunk 2]
```diff
@@ -33,6 +33,14 @@ import './p5.RenderBuffer';
 p5.RendererGL.prototype.beginShape = function(mode) {
   this.immediateMode.shapeMode =
     mode !== undefined ? mode : constants.TESS;
+  if (this._useUserAttributes === true){
+    for (const name of Object.keys(this.userAttributes)){
```
Comment (davepagurek) [inline src/webgl/p5.RendererGL.Immediate.js:37]: Is this equivalent to this? Might save making an intermediate array
```suggestion
    for (const name in this.userAttributes){
```

File: src/webgl/p5.RendererGL.Immediate.js [hunk 3]
```diff
@@ -113,6 +121,19 @@ p5.RendererGL.prototype.vertex = function(x, y) {
   const vert = new p5.Vector(x, y, z);
   this.immediateMode.geometry.vertices.push(vert);
   this.immediateMode.geometry.vertexNormals.push(this._currentNormal);
+  if (this._useUserAttributes){ 
+    const geom = this.immediateMode.geometry;
+    const verts = geom.vertices;
+    Object.entries(this.userAttributes).forEach(([name, data]) => {
+      const size = data.length ? data.length : 1;
+      if (verts.length > 0 && !geom.hasOwnProperty(name)) {
+        for (let i = 0; i < verts.length - 1; i++) {
```
Comment (davepagurek) [inline src/webgl/p5.RendererGL.Immediate.js:130]: Not urgent because this definitely works, but similar to what you did above in the geometry builder, you could maybe replace this for loop with something like `Array(numMissingValues).fill(0)`?

File: src/webgl/p5.RendererGL.Immediate.js [hunk 4]
```diff
@@ -165,6 +186,29 @@ p5.RendererGL.prototype.vertex = function(x, y) {
...
 
+p5.RendererGL.prototype.setAttribute = function(attributeName, data){
+  // if attributeName is in one of default, throw some warning
+  if(!this._useUserAttributes){
+    this._useUserAttributes = true;
+    this.userAttributes = {};
+  }
+  const size = data.length ? data.length : 1;
+  if (!this.userAttributes.hasOwnProperty(attributeName)){
+    this.tessyVertexSize += size;
```
Comment (davepagurek) [inline src/webgl/p5.RendererGL.Immediate.js:197]: nice!

File: src/webgl/p5.RendererGL.Immediate.js [hunk 5]
```diff
@@ -165,6 +186,29 @@ p5.RendererGL.prototype.vertex = function(x, y) {
...
+  }
+  this.userAttributes[attributeName] = data;
+  const buff = attributeName.concat('Buffer');
+  const bufferExists = this.immediateMode
+    .buffers
+    .user
+    .some(buffer => buffer.dst === buff);
```
Comment (davepagurek) [inline src/webgl/p5.RendererGL.Immediate.js:204]: No action needed here but just double check my assumptions: At first I was a tad worried that we're doing this loop for every vertex every frame, but I guess it only becomes a problem if you have a bunch of user attributes right? If you have one or two this is more than enough. Potentially if we wanted to scale this to more, we could refactor the buffers to be an object instead of an array, but that's maybe a bigger refactor than we need right now.

**Review by davepagurek — COMMENTED** [review]
Great work on this, especially getting the tessellation stuff to work!

File: src/webgl/p5.RendererGL.Immediate.js (hunk 3 above)
Comment (lukeplowden) [reply src/webgl/p5.RendererGL.Immediate.js]: I've just added a [size parameter](https://github.com/processing/p5.js/pull/7276/commits/6ec85c118074d4b03f56504ab2535b8c2f7d8c56 to `Geometry.setAttribute()` to make this easier.

)```js
setAttribute(attributeName, data, size = data.length ? data.length : 1){}
```

File: src/webgl/p5.RendererGL.Immediate.js (hunk 2 above)
Comment (lukeplowden) [reply src/webgl/p5.RendererGL.Immediate.js]: > Is this equivalent to this? Might save making an intermediate array

Thanks, I've replaced all references!

File: src/webgl/p5.RendererGL.Immediate.js (hunk 5 above)
Comment (lukeplowden) [reply src/webgl/p5.RendererGL.Immediate.js:205]: Yeah it seems to be okay for the moment, you can set quite a few attributes without it slowing down noticeably, but in the future we could look into that refactor.

File: src/webgl/p5.Geometry.js (hunk 1 above)
Comment (lukeplowden) [reply src/webgl/p5.Geometry.js:1922]: I will have to add something for this, at least to protect against `attributeName` being `aPosition`, `aNormal` and so on, but we can just concatenate something on the end of the javascript property, i.e. `attributeName` becomes `attributeNameSrc` for the first argument here:
```js
  if(!bufferExists){
    this.immediateMode.buffers.user.push(
      new p5.RenderBuffer(size, attributeName, buff, attributeName, this)
    );
  }  
```

File: src/core/shape/vertex.js [hunk 6]
```diff
@@ -2253,6 +2253,77 @@ p5.prototype.normal = function(x, y, z) {
...
 
+/** Sets the shader's vertex attribute variables.
+ * 
+ * Shader programs run on the computer's graphics processing unit (GPU)
+ * They live in a part of the computer's memory that's completely separate from
+ * the sketch that runs them. Attributes are variables attached to vertices
+ * within a shader program. They provide a way to attach data to vertices
+ * and pass values from a sketch running on the CPU to a shader program.
+ * 
+ * The first parameter, `attributeName`, is a string with the attribute's name.
```
Comment (davepagurek) [inline src/core/shape/vertex.js:2264]: Maybe worth a second sentence saying that it should match the name of the attribute in your shader, e.g. if you've got `attribute vec2 aOffset` (or `in vec2 aOffset` -- as an aside, do you think we should start using glsl es 300 more in the docs?), then you'd `setAttribute('aOffset', yourValue)`

File: src/core/shape/vertex.js [hunk 7]
```diff
@@ -2253,6 +2253,77 @@ p5.prototype.normal = function(x, y, z) {
   return this;
 };
 
+/** Sets the shader's vertex attribute variables.
+ * 
+ * Shader programs run on the computer's graphics processing unit (GPU)
+ * They live in a part of the computer's memory that's completely separate from
```
Comment (davepagurek) [inline src/core/shape/vertex.js:2259]: I wonder if this might be a bit more context than is needed at first. I'm planning to update the tutorials in the Learn section of the site when we do the 2.0 release to mention attributes and shader hooks, that might be a better place to explain the memory model? I wonder if all users need to know here is that we're adding extra data per vertex that you can read in a shader.

File: src/core/shape/vertex.js [hunk 8]
```diff
@@ -2253,6 +2253,77 @@ p5.prototype.normal = function(x, y, z) {
...
+ *  uniform mat4 uModelViewMatrix;
+ *  uniform mat4 uProjectionMatrix;
+ *  
+ *  attribute vec3 aPosition;
+ *  attribute vec2 aOffset;
+ * 
+ *  void main(){
+ *    vec4 positionVec4 = vec4(aPosition.xyz, 1.0);
+ *    positionVec4.xy += aOffset;   
+ *    gl_Position = uProjectionMatrix * uModelViewMatrix * positionVec4;
```
Comment (davepagurek) [inline src/core/shape/vertex.js:2286]: this is a great minimal example!

File: src/webgl/p5.Geometry.js (hunk 1 above)
Comment (davepagurek) [reply src/webgl/p5.Geometry.js:1923]: That's a great idea that doesn't involve shuffling around any of our other variables! Maybe just leave a comment somewhere that we can't name our builtin attributes with an Src suffix

File: src/webgl/p5.RendererGL.js [hunk 9]
```diff
@@ -2463,6 +2465,23 @@ p5.RendererGL = class RendererGL extends Renderer {
     return tessy;
   }
 
+  _updateTessyCombineCallback() {
+    // If custom attributes have been used, the vertex data which needs to be 
+    // combined has changed, so libtess must have its combine callback updated
+    const combinecallback = (coords, data, weight) => {
```
Comment (davepagurek) [inline src/webgl/p5.RendererGL.js:2471]: I'm just comparing this function to the initial combine callback we pass in, is the code the same or is there a part of it that's updated?
https://github.com/processing/p5.js/blob/8f463374cc33932f75931ac1a89e32ff63273559/src/webgl/p5.RendererGL.js#L2451-L2460

File: src/core/shape/vertex.js (hunk 6 above)
Comment (lukeplowden) [reply src/core/shape/vertex.js:2264]: I will add this

File: src/core/shape/vertex.js (hunk 7 above)
Comment (lukeplowden) [reply src/core/shape/vertex.js:2259]: Yeah you might be right, to be honest I copied it from [`setUniform()`](https://p5js.org/reference/p5.Shader/setUniform/) but they could both change.

File: src/webgl/p5.RendererGL.js (hunk 9 above)
Comment (lukeplowden) [reply src/webgl/p5.RendererGL.js:2471]: It is the same, it's just that the result array size is different:
```js
const result = new Array(p5.RendererGL.prototype.tessyVertexSize).fill(0); // original
const result = new Array(this.tessyVertexSize).fill(0);  // when it needs to be updated
```

File: src/core/shape/vertex.js (hunk 6 above)
Comment (lukeplowden) [reply src/core/shape/vertex.js:2264]: And yes, I definitely think we should move towards es 300 in the docs! I could mention one or both in this reference, and eventually update everything to `in`/`out` in documentation.

File: src/webgl/p5.RendererGL.js (hunk 9 above)
Comment (lukeplowden) [reply src/webgl/p5.RendererGL.js]: This has been changed now!

**Comment by lukeplowden** [comment]
- [Example 1](https://editor.p5js.org/lukeplowden/sketches/N65ukx26w)
- [Example 2](https://editor.p5js.org/lukeplowden/sketches/DGbl3OlNz)
- [Example 3 (on the Geometry method)](https://editor.p5js.org/lukeplowden/sketches/-yyQ9WYF9)

This is most of the way finished now! These are my examples for the function at the moment, feel free to try it and let me know what you think.

One consideration is that there is already an [XML  `setAttribute()`](https://p5js.org/reference/p5.XML/setAttribute/), a [`p5.Element.attribute()`](https://p5js.org/reference/p5.Element/attribute/), and a [WebGL canvas `setAttributes()`](https://p5js.org/reference/p5/setAttributes/). So now would be a good time to consider the name for this function. @limzykenneth, @davepagurek and I also wanted your thoughts on this, as you're apparently working on the canvas attributes function at the moment.

Also tagging @Garima3110 and @perminder-17 in case you have any thoughts or feedback in general.

---
### Reference:

Sets the shader's vertex attribute variables.

An attribute is a variable belonging to a vertex in a shader. p5.js provides some default attributes, such as `aPosition`, `aNormal`, `aVertexColor`, etc. These are set using `vertex()`, `normal()`, and `fill()` respectively. Custom attribute data can also be defined within `beginShape()` and `endShape()`.

The first parameter, `attributeName`, is a string with the attribute's name. This is the same variable name which should be declared in the shader, similar to `setUniform()`.

The second parameter, `data`, is the value assigned to the attribute. This value will be applied to subsequent vertices created with `vertex()`. It can be a Number or an array of numbers, and in the shader program the type can be declared according to the WebGL specification. Common types include `float`, `vec2`, `vec3`, `vec4` or matrices.

See also the `setAttribute()` method on Geometry objects.

File: src/core/shape/vertex.js [hunk 10]
```diff
@@ -2253,4 +2253,177 @@ p5.prototype.normal = function(x, y, z) {
...
+ *       vertex(x + cellSize, y + cellSize);
+ *       vertex(x, y + cellSize);
+ *     }
+ *   }
+ *   endShape();
+ * }
+ * </code>
+ * </div>
+/
+/**
```
Comment (davepagurek) [inline src/core/shape/vertex.js:2418]: I think we might need these to all be in one doc comment so that they don't show up as two separate overloads. Does it work if we remove this line and the one above?

File: src/webgl/p5.Geometry.js [hunk 11]
```diff
@@ -1910,6 +1919,119 @@ p5.Geometry = class Geometry {
...
+ *   rotateY(millis()*0.001);
+ *
+ *   // Draw the geometry
+ *   model(geo);
+ * }
+ * </code>
+ * </div>
+/
+/**
+ * @method setAttribute
```
Comment (davepagurek) [inline src/webgl/p5.Geometry.js:2015]: Same thing here for making it one doc comment, and also I think for methods on a class (as opposed to things we define via `p5.prototype.methodname = ...`), we don't need the `@method` tag any more.

**Review by davepagurek — COMMENTED** [review]
A thought on naming: would it be weird to not use `attribute` at all, and call it something like `setVertexData(name, value)`? (Or maybe `vertexData(...)` if we want to be consistent with `fill(...)` and `normal(...)`?) Or possibly `vertexProperty` to be clearer that it's adding something onto the vertex rather than being an alternative way to set a vertex?

File: src/webgl/p5.Geometry.js (hunk 11 above)
Comment (lukeplowden) [reply src/webgl/p5.Geometry.js]: Fixed these now :)

**Comment by lukeplowden** [comment]
> A thought on naming: would it be weird to not use `attribute` at all, and call it something like `setVertexData(name, value)`? (Or maybe `vertexData(...)` if we want to be consistent with `fill(...)` and `normal(...)`?) Or possibly `vertexProperty` to be clearer that it's adding something onto the vertex rather than being an alternative way to set a vertex?

Yeah I'm also happy with those, especially as `attribute` is replaced with `in` for 300 es. However `vertexAttribute()` or `vertexProperty()` are quite similar also.

File: src/webgl/p5.Geometry.js [hunk 12]
```diff
@@ -2010,11 +2010,10 @@ p5.Geometry = class Geometry {
  * }
  * </code>
  * </div>
-/
-/**
- * @method setAttribute
+ *
  * @param {String} attributeName the name of the vertex attribute.
  * @param {Number|Number[]} data the data tied to the vertex attribute.
+ * @param {Number} size optional size of each unit of data
```
Comment (davepagurek) [inline src/webgl/p5.Geometry.js:2016]: We can mark a parameter as optional in the friendly error system's validator too by putting its name in square brackets in the doc comments:

```suggestion
 * @param {Number} [size] optional size of each unit of data
```

**Comment by davepagurek** [comment]
I think this is looking good to go! @perminder-17 how do you feel about `vertexProperty` as a potential changed api name? if we all feel ok about that, I think we can rename it and then merge this!

**Comment by perminder-17** [comment]
Using the name `vertexProperty` makes it clear that we're adding something to the vertex, not replacing or setting the vertex itself. So, I think `vertexProperty` is a good name.

**Comment by davepagurek** [comment]
Thanks everyone!

**Comment by davepagurek** [comment]
@all-contributors please add @lukeplowden for code
